tests/test_Pretrain.py

```diff
@@ -1,41 +1,17 @@
 
 from __future__ import print_function
 import _setup_test_env  # noqa
-from nose.tools import assert_equal, assert_in, assert_not_in, assert_true
+from nose.tools import assert_equal, assert_in, assert_not_in
 from returnn.pretrain import pretrain_from_config
 from returnn.config import Config
 
-
-config1_dict = {
-  "pretrain": "default",
-  "num_inputs": 5,
-  "num_outputs": 10,
-  "hidden_size": (7, 8,),
-  "hidden_type": "forward",
-  "activation": "relu",
-  "bidirectional": False,
-}
-
-config2_dict = {
-  "use_tensorflow": True,
-  "pretrain": "default",
-  "num_inputs": 40,
-  "num_outputs": 4498,
-  "bidirectional": True,
-  "hidden_size": (500, 500, 500),
-  "hidden_type": "rec",
-  "activation": "sigmoid",
-  "dropout": 0.1,
-}
-
-config3_dict = {
-  "use_theano": True,
+config_dict = {
   "pretrain": "default",
   "num_inputs": 40,
   "num_outputs": 4498,
 }
 
-config3_json = """
+config_json = """
```

Comment:
How is this still relevant with `net_dict2`?

Reply:
It does trigger slightly different code. This can probably (did not check all details) be remove in the future as well (including the code related to this in config.py).

Reply:
But isn't it exactly the same? I don't see the difference.

Reply:
Also, the naming of these vars is confusing to me. Why call one `config_json`, and the other `net_dict`?

Reply:
Because one is a json string that gets parsed in the config.py via specifically setting the `network_topology_json`, the other is just our regular `net_dict`. I never knew this existed, and as I said we can probably remove it. Still, the tests are not doing exactly the same.

Reply:
Lets not discuss this, I will just remove it.

Reply:
But it should not be removed if there is a difference.

Reply:
Nothing that is directly relevant to the pre-training I think.

Reply:
But what exactly is different? I still don't understand this.

Reply:
`config.py 581 ff`
```python3
  if config.has("network") and config.is_typed("network"):
    json_content = config.typed_value("network")
    assert isinstance(json_content, dict)
    assert json_content
  elif config.network_topology_json:
    start_var = config.network_topology_json.find('(config:', 0)  # e.g. ..., "n_out" : (config:var), ...
    while start_var > 0:
      end_var = config.network_topology_json.find(')', start_var)
      assert end_var > 0, "invalid variable syntax at " + str(start_var)
      var = config.network_topology_json[start_var+8:end_var]
      assert config.has(var), "could not find variable " + var
      config.network_topology_json = (
        config.network_topology_json[:start_var] + config.value(var, "") + config.network_topology_json[end_var+1:])
      print("substituting variable %s with %s" % (var, config.value(var, "")), file=log.v4)
      start_var = config.network_topology_json.find('(config:', start_var+1)
    try:
      import json
      json_content = json.loads(config.network_topology_json)
    except ValueError as e:
      print("----- BEGIN JSON CONTENT -----", file=log.v3)
      print(config.network_topology_json, file=log.v3)
      print("------ END JSON CONTENT ------", file=log.v3)
      assert False, "invalid json content, %r" % e
    assert isinstance(json_content, dict)
    if 'network' in json_content:
      json_content = json_content['network']
    if not json_content:
      raise Exception("network json is empty: %s" % config.network_topology_json)
```
As I said before (and now verified) it is really just the json parsing. So it is different code, but not relevant to the pre-training, so I think the test can be removed and PROBABLY the whole code here as well.

Reply:
(In a future PR where we check this properly, that is)

Reply:
But as I said, JSON vs net dict, this is not a difference for this test.

I'm asking, if there is any real difference. Is there? If yes, then the test should stay (but remove the JSON aspect). And clarify what the difference is. If no, then remove it.

Reply:
I don't see the answer? So the answer is, "no, there is no difference"?

Reply:
It was the post before:

> As I said before (and now verified) it is really just the json parsing. So it is different code, but not relevant to the pre-training, so I think the test can be removed and PROBABLY the whole code here as well

Which says:
 - There is a difference with respect to loading the JSON vs. not loading the JSON
 - There is no difference with respect to the pre-training

Reply:
But as I said, the JSON aspect is not relevant. Let's not repeat that over and over.

I'm asking if there is any other difference. Not just w.r.t. pretraining, but potentially other things as well. I just want to know.